### ISSUE TITLE
Export/Import Settings

### ISSUE THREAD

**Issue by RickyM7** [florisboard/florisboard#467]
I think it would be very useful to have an option to export/import the settings of the app, after all there are many options and if there is a situation where it is necessary to uninstall and reinstall the app, do not need to configure everything from the beginning would be great.

**Comment by patrickgold** [comment]
Thanks for the suggestion! This is already planned and I could bet that it was already discussed somewhere, but can't find a separate issue for this, so leaving this one up.

**Comment by theotheroracle** [comment]
would it be possible to export themes and settings into one file that could be imported at once ? maybe just a zip archive with a metada file

**Comment by Luensche** [comment]
@patrickgold Do you have already ideas how to implement? Maybe I would have a try.

**Comment by patrickgold** [comment]
@Luensche The thing with Settings import/export is, that currently for a full backup you need to export quite some files, some are in xml, some in json, some a mysql db format (user dictionary). Additionally I want to rewrite the preferences backend shortly after v0.4.0, where the preference storage format will change. With this, I will then add pref i/e as a feature.

**Comment by RickyM7** [comment]
@patrickgold It might be interesting to look at how [NewPipe](https://github.com/TeamNewPipe/NewPipe) creates backups. The backup files generated by the NewPipe are a zip file with two different file types within it.

**Comment by just-Nob** [comment]
Happy to see that this feature is already planned! :-D
I'm able to "rescue" and backup settings by using [OAndBackupX](https://f-droid.org/de/packages/com.machiav3lli.backup/), but one the one hand not everybody has a rooted device, on the other hand [OAndBackupX](https://f-droid.org/de/packages/com.machiav3lli.backup/) won't help "syncing" my three FlorisBoard instances (FlorisBoard, FlorisBoard Beta, FlorisBoard Debug). ;-)

**Comment by D3SOX** [comment]
Participating in Android's backup system would be nice to allow e.g. Seedvault (which does not require root) to back them up
https://developer.android.com/guide/topics/data/backup
https://calyxinstitute.org/projects/seedvault-encrypted-backup-for-android

**Comment by just-Nob** [comment]
Thanks @D3SOX for your suggestion - SeedVault seems to be a nice tool!
Anyway, due to the fact that not all users caring about privacy are also able to compile APKs by themselves, I suggest to have the export/import functionality as an additional option.
I also guess that even SeedVault isn't able to transfer settings between FlorisBoard's different kinds of instances ('standard'/beta/debug).

**Comment by patrickgold** [comment]
>  "syncing" my three FlorisBoard instances (FlorisBoard, FlorisBoard Beta, FlorisBoard Debug)

Syncing these instances sounds really cool at first but actually it will make the older channel crash when the preference structure has changed a bit (and currently the subtype preference structure differs between stable and beta). Thus of course it can be useful but in some cases you end up making one channel unusable because the newer one requires the preferences to be different than the older one.

> Participating in Android's backup system would be nice to allow

Yes, I know. Currently backup support is disabled in the manifest, because I would have to properly configure the xml file. I may consider at least implementing the system backup support as a compromise for the longer wait on dedicated settings i/e.

**Comment by patrickgold** [comment]
Above PR implements manual backup and restore, which allows to fully back up app preferences, custom layouts, popup mappings, themes and spelling dictionaries. If you want to try out the backup and restore feature the PR's actions have the latest debug build attached to them. This content will be released in beta09 when some other features for beta09 are finished.
